### ISSUE TITLE
A rule to prevent people from changing the `testRegex`/`testMatch`

### ISSUE THREAD

**Issue by kasir-barati** [jest-community/eslint-plugin-jest#1785]
Hi,

I almost all the times accidentally change these two (of course for testing and getting quick feedbacks), so I was hoping to have a rule to setup in my eslint config file so that I cannot change them to a certain file and commit/merge/deploy my code.

Something like this:

```js
rules: {
  'jest/test-regex': '.e2e-spec.ts$',
}
```

Which means that my jest conf file should look like this:

```json
{
  "testRegex": ".e2e-spec.ts$",
}
```

I mean if I go and change it to `a.e2e-spec.ts` it should complain. Same thing for `testMatch`. is this possible with this lib? If not, do you think it worth adding?

**Comment by G-Rath** [comment]
hmm I feel like there are better alternatives for this kind of problem, such as:
  * using [`husky`](https://www.npmjs.com/package/husky) and [`lint-staged`](https://github.com/lint-staged/lint-staged) to do precommit checks
  * using [`--testPathPatterns`](https://jestjs.io/docs/cli#--testpathpatternsregex)
  * having checks in your actual Jest config that get triggered by env variables

That last one is an easily forgotten about "feature" - because you can author your config in JavaScript you can also do checks and logic e.g.

```js
// jest.config.js
const defaultTestRegex = /.e2e-spec.ts$/u;

const config = {
  testRegex: defaultTestRegex,
};

// this check ensures the default regexp gets updated as part of your CI
if (process.env.CI && config.testRegex.source !== defaultTestRegex.source) {
  throw new Error('test regex has been changed without updating the default!');
}

// or maybe you could override it from an external env variable, which works well with tools like https://direnv.net/
if (process.env.LOCAL_JEST_TEST_REGEXP) {
  config.testRegex = new RegExp(process.env.LOCAL_JEST_TEST_REGEXP);
  console.warn(
    'test regex has been replaced with',
    process.env.LOCAL_JEST_TEST_REGEXP,
    '- remember that only some tests will be run!',
  );
}

module.exports = config;
```

**Comment by kasir-barati** [comment]
Thanks @G-Rath this helped me a lot. I had to convert my `jest.config.json` file to a JS/TS file though.

**Comment by kasir-barati** [comment]
Now that I think about it I guess that* was the reason why I did not think of coding it.

*The fact that my jest config file was a JSON file. Though it is not always that simple, take for instance when you've all of your jest configurations in your `package.json` file. Moving them outside of it might introduce additional changes. But I guess those changes are fine.

**Comment by SimenB** [comment]
Linting config in general might be a nice thing - e.g. warning about changing defaults or some such between majors.

**Comment by kasir-barati** [comment]
@SimenB @G-Rath Should we change the title of this issue to "Linting configs" an reopen it?

**Comment by G-Rath** [comment]
I don't really feel like a linting rule for configs is that useful because imo Jest itself should be telling people about deprecations and upcoming major config changes.

Linting rules to me are better for stuff that can understandable come up repeatedly (even if just locally) e.g. no alias matches was for code that could be written in future, whereas configs tend to be mostly "write once" and not have many gotchas.

Happy to be convinced otherwise though if folks have examples of where a lint rule would have been better 🤷‍♂️
